File: CyxbsMobile2019_iOS/Rising Schedule/ScheduleCollectionViewCell.m
//
//  ScheduleCollectionViewCell.m
//  CyxbsMobile2019_iOS
//
//  Created by SSR on 2022/9/5.
//  Copyright © 2022 Redrock. All rights reserved.
//

#import "ScheduleCollectionViewCell.h"

NSString *ScheduleCollectionViewCellReuseIdentifier = @"ScheduleCollectionViewCell";

#pragma mark - ScheduleCollectionViewCell ()

@interface ScheduleCollectionViewCell ()

/// 标题
@property (nonatomic, strong) UILabel *titleLab;

/// 细节
@property (nonatomic, strong) UILabel *contentLab;

/// 多人
@property (nonatomic, strong) UIView *multyView;

@end

#pragma mark - ScheduleCollectionViewCell

@implementation ScheduleCollectionViewCell

#pragma mark - Life cycle

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        _drawType = -1;
        self.backgroundColor = UIColor.clearColor;
        self.contentView.layer.cornerRadius = 8;
        self.contentView.clipsToBounds = YES;
        [self.contentView addSubview:self.multyView];
        [self.contentView addSubview:self.titleLab];
        [self.contentView addSubview:self.contentLab];
    }
    return self;
}

#pragma mark - Method

- (void)applyLayoutAttributes:(UICollectionViewLayoutAttributes *)layoutAttributes {
    CGRect frame = layoutAttributes.frame;
    
    self.multyView.right = frame.size.width - 5;
    
    self.titleLab.width = frame.size.width - 2 * self.titleLab.left;
    
    self.contentLab.left = self.titleLab.left;
    self.contentLab.width = self.titleLab.width;
    self.contentLab.bottom = frame.size.height - 8;
}

#pragma mark - Getter

- (UILabel *)titleLab {
    if (_titleLab == nil) {
        _titleLab = [[UILabel alloc] initWithFrame:CGRectMake(8, 8, -1, -1)];
        _titleLab.backgroundColor = UIColor.clearColor;
        _titleLab.numberOfLines = 3;
        _titleLab.font = [UIFont fontWithName:PingFangSC size:10];
        _titleLab.textAlignment = NSTextAlignmentCenter;
    }
    return _titleLab;
}

- (UILabel *)contentLab {
    if (_contentLab == nil) {
        _contentLab = [[UILabel alloc] init];
        _contentLab.backgroundColor = UIColor.clearColor;
        _contentLab.numberOfLines = 3;
        _contentLab.font = [UIFont fontWithName:PingFangSC size:10];
        _contentLab.textAlignment = NSTextAlignmentCenter;
    }
    return _contentLab;
}

- (UIView *)multyView {
    if (_multyView == nil) {
        _multyView = [[UIView alloc] initWithFrame:CGRectMake(-1, 4, 8, 2)];
        _multyView.layer.cornerRadius = _multyView.height / 2;
        _multyView.clipsToBounds = YES;
        _multyView.alpha = 0;
    }
    return _multyView;
}

- (NSString *)courseTitle {
    return self.titleLab.text.copy;
}

- (NSString *)courseContent {
    return self.contentLab.text.copy;
}

#pragma mark - Setter

- (void)setCourseTitle:(NSString *)courseTitle {
    self.titleLab.text = courseTitle.copy;
    CGFloat width = self.titleLab.width;
    [self.titleLab sizeToFit];
    self.titleLab.width = width;
}

- (void)setCourseContent:(NSString *)courseContent {
    self.contentLab.text = courseContent.copy;
    CGFloat width = self.contentLab.width;
    CGFloat bottom = self.contentLab.bottom;
    [self.contentLab sizeToFit];
    self.contentLab.width = width;
    self.contentLab.bottom = bottom;
}

- (void)setMultipleSign:(BOOL)multipleSign {
    if (_multipleSign != multipleSign) {
        return;
    }
    _multipleSign = multipleSign;
    
    self.multyView.alpha = multipleSign ? 1 : 0;
}

- (void)setDrawType:(ScheduleCollectionViewCellDrawType)drawType {
    if (_drawType == drawType) {
        return;
    }
    _drawType = drawType;
    
    switch (drawType) {
        /*ScheduleCollectionViewCellDrawMorning*/
        case ScheduleCollectionViewCellDrawMorning: {
            self.contentView.backgroundColor =
            [UIColor dm_colorWithLightColor:UIColorHex(#F9E7D8)
                                  darkColor:UIColorHex(#FFCCA126)];
            
            self.titleLab.textColor = self.contentLab.textColor = self.multyView.backgroundColor =
            [UIColor dm_colorWithLightColor:UIColorHex(#FF8015)
                                  darkColor:UIColorHex(#F0F0F2CC)];
        } break;
            
        /*ScheduleCollectionViewCellDrawAfternoon*/
        case ScheduleCollectionViewCellDrawAfternoon: {
            self.contentView.backgroundColor =
            [UIColor dm_colorWithLightColor:UIColorHex(#F9E3E4)
                                  darkColor:UIColorHex(#FF979B26)];
            
            self.titleLab.textColor = self.contentLab.textColor = self.multyView.backgroundColor =
            [UIColor dm_colorWithLightColor:UIColorHex(#FF6262)
                                  darkColor:UIColorHex(#F0F0F2CC)];
        } break;
            
        /*ScheduleCollectionViewCellDrawNight*/
        case ScheduleCollectionViewCellDrawNight: {
            self.contentView.backgroundColor =
            [UIColor dm_colorWithLightColor:UIColorHex(#DDE3F8)
                                  darkColor:UIColorHex(#9BB2FB26)];
            
            self.titleLab.textColor = self.contentLab.textColor = self.multyView.backgroundColor =
            [UIColor dm_colorWithLightColor:UIColorHex(#4066EA)
                                  darkColor:UIColorHex(#F0F0F2CC)];
        } break;
            
        /*ScheduleCollectionViewCellDrawOthers*/
        case ScheduleCollectionViewCellDrawOthers: {
            self.contentView.backgroundColor =
            [UIColor dm_colorWithLightColor:UIColorHex(#DFF3FC)
                                  darkColor:UIColorHex(#90DBFB26)];
            
            self.titleLab.textColor = self.contentLab.textColor = self.multyView.backgroundColor =
            [UIColor dm_colorWithLightColor:UIColorHex(#06A3FC)
                                  darkColor:UIColorHex(#F0F0F2CC)];
        } break;
            
        /*ScheduleCollectionViewCellDrawCustom*/
        case ScheduleCollectionViewCellDrawCustom: {
            self.contentView.backgroundColor =
            [UIColor dm_colorWithLightColor:UIColorHex(#E9ECF0)
                                  darkColor:UIColorHex(#3C4E70)];
            
            self.titleLab.textColor = self.contentLab.textColor = self.multyView.backgroundColor =
            [UIColor dm_colorWithLightColor:UIColorHex(#4D4D4D)
                                  darkColor:UIColorHex(#F0F0F2CC)];
        } break;
    }
}

@end

File: CyxbsMobile2019_iOS/Discover/Discover_WeDate/CourseSchedule/ScheduleCollectionViewCell.swift
//
//  ScheduleCollectionViewCell.swift
//  CyxbsMobile2019_iOS
//
//  Created by coin on 2023/9/17.
//  Copyright © 2023 Redrock. All rights reserved.
//

import UIKit

class ScheduleCollectionViewCell: UICollectionViewCell {
    
    // MARK: ReuseIdentifier
    
    static let curriculumReuseIdentifier = "CyxbsMobile2019_iOS.ScheduleCollectionViewCell.curriculum"
    static let supplementaryReuseIdentifier = "CyxbsMobile2019_iOS.ScheduleCollectionViewCell.supplementary"
    
    // MARK: DrawType
    
    enum DrawType {
        
        enum CurriculumType {
            case allBusy
            case leisureMore
            case busyMore
            case allLeisure
        }
        
        enum SupplementaryType {
            case normal
            case today
        }
        
        case curriculum(CurriculumType)
        case supplementary(SupplementaryType)
    }
    
    // MARK: init

    override init(frame: CGRect) {
        super.init(frame: frame)
        backgroundColor = .clear
        contentView.layer.cornerRadius = 8
        contentView.clipsToBounds = true
        contentView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        
        contentView.addSubview(titleLab)
        contentView.addSubview(contentLab)
        updateFrame()
        
        contentView.insertSubview(darkBackView, at: 0)
        contentView.insertSubview(lightBackView, at: 0)
        darkBackView.isHidden = true
        lightBackView.isHidden = true
        
        contentView.addSubview(timePointerView)
        timePointerView.isHidden = true
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    private(set) var drawType: DrawType = .curriculum(.allBusy)
    
    private(set) var isTitleOnly: Bool = false
    
    override func apply(_ layoutAttributes: UICollectionViewLayoutAttributes) {
        super.apply(layoutAttributes)
        updateFrame()
    }
    
    override func layoutSubviews() {
        super.layoutSubviews()
        lightBackView.frame = contentView.frame
        lightBackView.top += 10
        darkBackView.frame = contentView.frame
        darkBackView.height -= 5
        timePointerView.frame = CGRect(x: 2, y: 0, width: contentView.width - 3, height: 6)
    }
    
    // MARK: lazy
    
    lazy var titleLab: UILabel = {
        let lab = UILabel()
        lab.backgroundColor = .clear
        lab.textAlignment = .center
        lab.numberOfLines = 3
        return lab
    }()
    
    lazy var contentLab: UILabel = {
        let lab = UILabel()
        lab.backgroundColor = .clear
        lab.textAlignment = .center
        lab.numberOfLines = 3
        return lab
    }()
    
    private lazy var lightBackView: UIView = {
        let lightBackView = UIView()
        lightBackView.backgroundColor = UIColor(hexString: "#F3F6FD", alpha: 1)
        return lightBackView
    }()
    
    private lazy var darkBackView: UIView = {
        let darkBackView = UIView()
        darkBackView.backgroundColor = UIColor(hexString: "#2A4E84", alpha: 1)
        darkBackView.layer.cornerRadius = 8
        darkBackView.clipsToBounds = true
        return darkBackView
    }()
    
    lazy var timePointerView: UIImageView = {
        let timePointerView = UIImageView()
        timePointerView.image = UIImage(named: "课表指示")
        return timePointerView
    }()
}

// MARK: - init

extension ScheduleCollectionViewCell {
    
    func initCurriculum() {
        drawType = .curriculum(.allBusy)
        titleLab.font = .systemFont(ofSize: 10, weight: .regular)
        contentLab.font = .systemFont(ofSize: 10, weight: .regular)
    }
    
    func initSupplementary() {
        drawType = .supplementary(.normal)
        titleLab.font = .systemFont(ofSize: 12, weight: .regular)
        contentLab.font = .systemFont(ofSize: 11, weight: .regular)
    }
}

// MARK: Method

extension ScheduleCollectionViewCell {
    
    func set(curriculumType: DrawType.CurriculumType, title: String?) {
        drawType = .curriculum(curriculumType)
        initCurriculum()
        titleLab.text = title
        switch curriculumType {
            
        case .allBusy:
            contentView.backgroundColor = UIColor(hexString: "#F9E3E4", alpha: 1)
            titleLab.textColor = UIColor(hexString: "#FF6262", alpha: 1)
            
        case .leisureMore:
            contentView.backgroundColor = UIColor(hexString: "#DDE3F8", alpha: 1)
            titleLab.textColor = UIColor(hexString: "#4066EA", alpha: 1)
            
        case .busyMore:
            contentView.backgroundColor = UIColor(hexString: "#F9E7D8", alpha: 1)
            titleLab.textColor = UIColor(hexString: "#FF8015", alpha: 1)
            
        case .allLeisure:
            contentView.backgroundColor = .clear
            titleLab.textColor = .clear
        }
        updateFrame()
    }
    
    func set(supplementaryType: DrawType.SupplementaryType, title: String?, content: String?, isTitleOnly: Bool, isNeedTimePointer: Bool) {
        drawType = .supplementary(supplementaryType)
        initSupplementary()
        self.isTitleOnly = isTitleOnly
        titleLab.text = title
        contentLab.text = content
        switch supplementaryType {
            
        case .normal:
            contentView.backgroundColor = .white
            contentView.layer.cornerRadius = 0
            titleLab.textColor = UIColor(hexString: "#15315B", alpha: 1)
            contentLab.textColor = UIColor(hexString: "#606E8A", alpha: 1)
            lightBackView.isHidden = true
            darkBackView.isHidden = true
            
            if isNeedTimePointer {
                timePointerView.isHidden = false
            } else {
                timePointerView.isHidden = true
            }
            
        case .today:
            titleLab.textColor = .white
            contentView.backgroundColor = .white
            lightBackView.isHidden = false
            darkBackView.isHidden = false
        }
        updateFrame()
    }
    
    func updateFrame() {
        switch drawType {
        case .curriculum(_):
            titleLab.frame.size.width = bounds.width - 2 * 13
            titleLab.sizeToFit()
            titleLab.center.x = bounds.width / 2
            titleLab.center.y = bounds.height / 2
            contentLab.isHidden = true
            
        case .supplementary(_):
            titleLab.sizeToFit()
            titleLab.frame.origin.x = 0
            titleLab.frame.size.width = bounds.width
            
            contentLab.sizeToFit()
            contentLab.frame.origin.x = 0
            contentLab.frame.size.width = bounds.width
            
            if isTitleOnly {
                titleLab.center.y = bounds.height / 2
            } else {
                titleLab.frame.origin.y = 13
            }
            contentLab.isHidden = isTitleOnly
            contentLab.frame.origin.y = bounds.height - contentLab.bounds.height - 13
        }
    }
}
